src/mame/misc/anes.cpp

```diff
@@ -38,149 +39,597 @@ namespace {
+	{
+		for (int x = 0; x < sw; x++)
+		{
+			for (int y = 0; y < sh; y++)
+			{
+				int drawx, drawy;
+
+				if (!flipy)
+					drawy = ((sy + y) & 0x1ff);
+				else
+					drawy = ((sy + (sh - 1) - y) & 0x1ff);
+
+				if (!flipx)
+					drawx = ((sx + x) & 0x1ff);
+				else
+					drawx = ((sx + (sw - 1) - x) & 0x1ff);
+
+				int pen = (which & 0x10) ? 0 : m_blitrom[(addr++) & (m_blitrom.bytes() - 1)];
+
+				if (pen != 0xff)
+					bitmap.pix(drawy, drawx) = pen;
+				else
+					bitmap.pix(drawy, drawx) = 0x00;
+			}
+		}
+	}
+	else
+	{
+		for (int y = 0; y < sh; y++)
+		{
+			for (int x = 0; x < sw; x++)
+			{
+				int drawx, drawy;
+
+				if (!flipy)
+					drawy = ((sy + y) & 0x1ff);
+				else
+					drawy = ((sy + (sh - 1) - y) & 0x1ff);
+
+				if (!flipx)
+					drawx = ((sx + x) & 0x1ff);
+				else
+					drawx = ((sx + (sw - 1) - x) & 0x1ff);
+
+				int pen = (which & 0x10) ? 0 : m_blitrom[(addr++) & (m_blitrom.bytes() - 1)];
+
+				if (pen != 0xff)
+					bitmap.pix(drawy, drawx) = pen;
+				else
+					bitmap.pix(drawy, drawx) = 0x00;
+			}
+		}
+	}
 }
 
-void anes_state::blit_trigger_w(uint8_t data)
+void anes_state::rombank_w(uint8_t data)
 {
-	/*
-	 operation is:
-	 checks for bit 4 to be high in port $16
-	 writes either 0 or 3 to port $04
-	 writes a 0 to port $00
-	 writes an offset to ports $0c / $0d / $0e
-	 writes 1 to port $0b, writes to program space, writes 2 to port $0b, writes to program space
-	 writes a mode to port $0b, writes to trigger port $0a
-	 */
+	if (data & (~0x07))
+		logerror("%s: Unknown ROM bank bit written %02X\n", machine().describe_context(), data);
+
+	m_bank = data & 0x07;
+
+	// the bank change must happen AFTER the next instruction, be it a jmp or a ret
+	m_bank_delay = 2;
+}
 
-	//printf("%02x%02x%02x\n",m_vram_offset[0],m_vram_offset[1],m_vram_offset[2]);
+void anes_state::blit_w(offs_t offset, uint8_t data)
+{
+	m_blit[offset] = data;
+	if (offset == 0x0a)
+		do_blit();
+}
+
+void anes_state::blit_rom_w(offs_t offset, uint8_t data)
+{
+	uint8_t which = m_blit[0x0b];
+//  logerror("%s: blit ROM written %02X (%04X <- %02X)\n", machine().describe_context(), which, offset, data);
+	if (which == 1)
+	{
+		m_blit_addr[0] = offset;
+		m_blit_val[0] = data;
+	}
+	else if (which == 2)
+	{
+		m_blit_addr[1] = offset;
+		m_blit_val[1] = data;
+	}
+	else
+		logerror("%s: Unknown blit ROM which %02X\n", machine().describe_context(), which);
+}
+
+uint8_t anes_state::blit_status_r()
+{
+	// bits 0-3 = ?
+	// bit    4 = busy
+	return 0x00;
+}
+
+void anes_state::mux_w(uint8_t data)
+{
+	m_mux = data;
+	if (m_mux & (~0x1f))
+		logerror("%s: Unknown mux bit written %02X\n", machine().describe_context(), data);
+}
+
+uint8_t anes_state::key_r(offs_t offset)
+{
+	for (int i = 0; i < 5; i++)
+		if (!BIT(m_mux, i))
+			return m_key[offset][i]->read();
+
+	//  logerror("%s: Unknown key read with mux = %02X\n", machine().describe_context(), m_mux);
+	//  return 0xff;
+	return offset ? m_coin->read() : 0xff;
+}
```

Comment:
This is a fairly common mahjong key matrix, not a real multiplexer (i.e. the mux_w fn name is misleading).

Reply:
this was from the code I was sent